### ISSUE TITLE
ActionCache: explicit delete operation?

### ISSUE THREAD

**Issue by ulfjack** [bazelbuild/remote-apis#138]
The action cache currently does *not* have a delete operation. According to GCP, the RBE permissions include a delete permission. One could consider an update request without a result as a delete. However, this isn't clear from the documentation. I think I'd have a slight preference for having a separate delete operation. WDYT?

**Comment by werkt** [comment]
"Without a result" presumably meaning the default `ActionResult` instance? Hard to make a request otherwise.

Incidentally, can you refer to the documentation where this permission is enumerated here. I went looking for something similar recently, and I think we need the language associated with the permission to agree on action anyway.

**Comment by EricBurnett** [comment]
I don't recall why RBE has an action 'delete' permission - most likely the permission exists but with no associated methods that leverage it. What is your use-case that needs Delete?

Whatever the reason, I have no specific objection to a Delete* method for actions, if you have a reason to need such a thing. But I also haven't encountered much call for it, and will note that RBE has no such need (or method) today*.

_* Through the front door, anyways - we remove AC entries with garbage collection, for example, but that's an implementation detail._

**Comment by ulfjack** [comment]
Google's RBE permissions are documented here: https://cloud.google.com/iam/docs/permissions-reference (search for remotebuildexecution)

@werkt: the Java protobuf API does distinguish between unset protos and default protos, it only doesn't allow doing this for primitive types.

We may need an API to delete malformed, broken, or malicious action cache entries. Technically, overwriting an action cache entry may also be possible by setting `skip_cache_lookup=true`, but that incurs the cost of re-executing the action, and the service typically only writes an entry if the exit code is zero.

**Comment by bergsieker** [comment]
Conceptually I have no objection to an explicit delete API. We originally opted not to implement one at least in part because constructing the digest to delete is relatively difficult, but that was purely a pragmatic decision based on the client tooling available at the time.

**Comment by peterebden** [comment]
I'd vote in favour of an explicit delete in preference to overloading UpdateActionResult. Agreed that this is a useful thing in practice.

We might want to consider extending ActionCacheUpdateCapabilities to indicate whether delete is enabled (or clarify that `update_enabled` covers both).

**Comment by sstriker** [comment]
In concept I would agree.  @ulfjack, would you mind bringing this back to the mailing list to ensure broader visibility of the question?  If you already have a proposal for the DeleteActionResult API, maybe open a WIP PR?

**Comment by sluongng** [comment]
This is rather an old issue, but I ran into this a while ago, so I wanted to comment on how this could be solved.

Since most AC server implementations check for the existence of output blobs in the CAS before returning the ActionResult, I would say the way to issue a deletion would be to update the same AC key with an ActionResult pointing to an invalid digest (i.e., a randomized blob's SHA-256 that does not exist in CAS) as the sole output file.

**Comment by peterebden** [comment]
Yeah, this is an old one - I had forgotten until it came up in my notifications!

I will say that we saw need for this sort of thing a few times (in my past life now): if you end up with a 'corrupt' AR (e.g. one that references blobs that don't exist) then you really do want that thing excised. That's a fairly easy state to get to if you have an inexact GC type thing going on (and it's remarkably hard to do that correctly), but we also had suspicion of one case where the evidence seemed to point to our CSP losing a blob - history will never be able to relate precisely what happened there, but we had to drop the associated AR pretty urgently. That was a manual operation but an API for it would have been welcome.

> Since most AC server implementations check for the existence of output blobs in the CAS before returning the ActionResult, I would say the way to issue a deletion would be to update the same AC key with an ActionResult pointing to an invalid digest (i.e., a randomized blob's SHA-256 that does not exist in CAS) as the sole output file.

I don't think I agree with this. That assumes that the server follows that behaviour, which I don't recall is actually required by REAPI; it also seems very reasonable for a server to reject such a request saying the referenced action result doesn't exist.

I'd prefer issuing a request to set the digest to that of the empty blob, which at least can be cleanly special-cased. Although I would still more prefer (prefer more?) adding an explicit extra method; there's not any real overhead to having an extra service method in gRPC and it is cleaner (and would map to a REST API nicely, if someone wanted to do that).

There is also a question of whether any REAPI client should be able to do this thing. There's no 'official' permissioning within the API, but to me this seems like more of an 'admin' operation, which is an argument for it existing outside REAPI.

**Comment by EdSchouten** [comment]
Why can't we extend the description of UpdateActionResult to say that if UpdateActionResult.action_result is unset, that it requests a delete?

**Comment by peterebden** [comment]
> Why can't we extend the description of UpdateActionResult to say that if UpdateActionResult.action_result is unset, that it requests a delete?

That also seems fine to me. I mildly prefer the more explicit delete RPC, but I'd not be unhappy with this.
